File: src/main/java/uss/code/course/dto/response/CourseResponse.java
package uss.code.course.dto.response;

import lombok.Builder;
import uss.code.course.domain.Course;
import uss.code.course.dto.common.CachedCourse;
import uss.code.course.dto.common.CourseCapacity;

import java.util.List;

import static lombok.AccessLevel.PRIVATE;

@Builder(access = PRIVATE)
public record CourseResponse(
        String id,

        String code,

        String courseCode,

        String name,

        String nameEn,

        String englishCourseName,

        String professor,

        int credits,

        int capacity,

        int enrolled,

        int cartCount,

        String courseType,

        String courseArea,

        String department,

        String grade,

        String schedule,

        List<String> tags,

        boolean isEnglish,

        boolean isNight,

        boolean isClosed
) {
    public static CourseResponse from(final Course course) {
        return CourseResponse.of(
                CachedCourse.from(course),
                course.getMaxCapacity(),
                course.getCurrentEnrollment(),
                course.getCartCount(),
                !(course.isActive() && course.isRegisterable())
        );
    }

    public static CourseResponse of(
            final CachedCourse course,
            final CourseCapacity capacity
    ) {
        return CourseResponse.of(
                course,
                capacity.maxCapacity(),
                capacity.currentEnrollment(),
                capacity.cartCount(),
                !capacity.isRegisterable()
        );
    }

    private static CourseResponse of(
            final CachedCourse course,
            final int capacity,
            final int enrolled,
            final int cartCount,
            final boolean isClosed
    ) {
        return CourseResponse.builder()
                .id(String.valueOf(course.id()))
                .code(course.code())
                .courseCode(course.courseCode())
                .name(course.name())
                .nameEn(course.nameEn())
                .englishCourseName(course.englishCourseName())
                .professor(course.professor())
                .credits(course.credits())
                .capacity(capacity)
                .enrolled(enrolled)
                .cartCount(cartCount)
                .courseType(course.courseType())
                .courseArea(course.courseArea())
                .department(course.department())
                .grade(course.grade())
                .schedule(course.schedule())
                .tags(course.tags())
                .isEnglish(course.isEnglish())
                .isNight(course.isNight())
                .isClosed(isClosed)
                .build();
    }
}

File: src/main/kotlin/uss/code/course/dto/response/CourseResponse.kt
package uss.code.course.dto.response

import uss.code.course.domain.Course
import uss.code.course.dto.internal.CachedCourseDto
import uss.code.course.dto.internal.CourseCapacityDto

@JvmRecord
data class CourseResponse(
    val id: String,

    val code: String,

    val courseCode: String,

    val name: String,

    val nameEn: String,

    val englishCourseName: String,

    val professor: String,

    val credits: Int,

    val capacity: Int,

    val enrolled: Int,

    val cartCount: Int,

    val courseType: String,

    val courseArea: String,

    val department: String,

    val grade: String,

    val schedule: String,

    val tags: List<String>,

    val isEnglish: Boolean,

    val isNight: Boolean,

    val isClosed: Boolean,
) {
    companion object {
        @JvmStatic
        fun of(
            course: CachedCourseDto,
            capacity: CourseCapacityDto,
        ): CourseResponse = of(
            course = course,
            capacity = capacity.maxCapacity,
            enrolled = capacity.currentEnrollment,
            cartCount = capacity.cartCount,
            isClosed = !capacity.isRegisterable(),
        )

        @JvmStatic
        fun from(course: Course): CourseResponse = of(
            course = CachedCourseDto.from(course),
            capacity = course.maxCapacity,
            enrolled = course.currentEnrollment,
            cartCount = course.cartCount,
            isClosed = !(course.isActive() && course.isRegisterable()),
        )

        private fun of(
            course: CachedCourseDto,
            capacity: Int,
            enrolled: Int,
            cartCount: Int,
            isClosed: Boolean,
        ): CourseResponse = CourseResponse(
            id = course.id.toString(),
            code = course.code,
            courseCode = course.courseCode,
            name = course.name,
            nameEn = course.nameEn,
            englishCourseName = course.englishCourseName,
            professor = course.professor,
            credits = course.credits,
            capacity = capacity,
            enrolled = enrolled,
            cartCount = cartCount,
            courseType = course.courseType,
            courseArea = course.courseArea,
            department = course.department,
            grade = course.grade,
            schedule = course.schedule,
            tags = course.tags,
            isEnglish = course.isEnglish,
            isNight = course.isNight,
            isClosed = isClosed,
        )
    }
}
